File: apps/api/internal/models/paper.go
package models

import "time"

type PaperAccount struct {
	UserID       string    `json:"user_id"`
	InitialCash  float64   `json:"initial_cash"`
	CashBalance  float64   `json:"cash_balance"`
	BaseCurrency string    `json:"base_currency"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

type PaperHolding struct {
	ID             string    `json:"id"`
	UserID         string    `json:"user_id"`
	InstrumentID   string    `json:"instrument_id"`
	Symbol         string    `json:"symbol,omitempty"`
	Name           string    `json:"name,omitempty"`
	Currency       string    `json:"currency,omitempty"`
	Quantity       float64   `json:"quantity"`
	AvgCost        float64   `json:"avg_cost"`
	CurrentPrice   float64   `json:"current_price,omitempty"`
	MarketValue    float64   `json:"market_value,omitempty"`
	MarketValueKRW float64   `json:"market_value_krw,omitempty"`
	PnLKRW         float64   `json:"pnl_krw,omitempty"`
	PnLPct         float64   `json:"pnl_pct,omitempty"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

type PaperTransaction struct {
	ID           string    `json:"id"`
	UserID       string    `json:"user_id"`
	InstrumentID string    `json:"instrument_id"`
	Symbol       string    `json:"symbol,omitempty"`
	Action       string    `json:"action"` // "buy" | "sell"
	Quantity     float64   `json:"quantity"`
	Price        float64   `json:"price"`
	Currency     string    `json:"currency"`
	FxToKRW      float64   `json:"fx_to_krw"`
	TotalKRW     float64   `json:"total_krw"`
	Active       bool      `json:"active"`
	CreatedAt    time.Time `json:"created_at"`
}

// EquityPoint — 일자별 KRW 평가액 (시계열용)
type EquityPoint struct {
	Date      string  `json:"date"`       // YYYY-MM-DD
	EquityKRW float64 `json:"equity_krw"`
}

// PaperTransactionCreate — POST body
type PaperTransactionCreate struct {
	InstrumentID string  `json:"instrument_id"`
	Action       string  `json:"action"`
	Quantity     float64 `json:"quantity"`
	Reason       *string `json:"reason,omitempty"`
}

// PaperPortfolioResponse — GET /v1/paper/portfolio 응답
type PaperPortfolioResponse struct {
	Account      PaperAccount  `json:"account"`
	Holdings     []PaperHolding `json:"holdings"`
	Summary      PaperSummary  `json:"summary"`
	EquitySeries []EquityPoint `json:"equity_series"`
}

type PaperSummary struct {
	TotalEquityKRW float64 `json:"total_equity_krw"`
	TotalPnLKRW    float64 `json:"total_pnl_krw"`
	TotalPnLPct    float64 `json:"total_pnl_pct"`
}

// PaperResetRequest — POST /v1/paper/reset body
type PaperResetRequest struct {
	InitialCash *float64 `json:"initial_cash,omitempty"`
}

File: apps/web/lib/api/paper.ts
import { authFetch } from "./auth-fetch";

export type PaperAccount = {
  user_id: string;
  initial_cash: number;
  cash_balance: number;
  base_currency: "KRW";
  created_at: string;
  updated_at: string;
};

export type PaperHolding = {
  id: string;
  user_id: string;
  instrument_id: string;
  symbol: string;
  name: string;
  currency: string;
  quantity: number;
  avg_cost: number;
  current_price: number;
  market_value: number;
  market_value_krw: number;
  pnl_krw: number;
  pnl_pct: number;
  created_at: string;
  updated_at: string;
};

export type PaperTransaction = {
  id: string;
  user_id: string;
  instrument_id: string;
  symbol: string;
  action: "buy" | "sell";
  quantity: number;
  price: number;
  currency: string;
  fx_to_krw: number;
  total_krw: number;
  active: boolean;
  created_at: string;
};

export type EquityPoint = { date: string; equity_krw: number };

export type PaperPortfolioResponse = {
  account: PaperAccount;
  holdings: PaperHolding[];
  summary: {
    total_equity_krw: number;
    total_pnl_krw: number;
    total_pnl_pct: number;
  };
  equity_series: EquityPoint[];
};

export type TradeError = {
  code: string;
  message: string;
  need_krw?: number;
  have_krw?: number;
  need_qty?: number;
  have_qty?: number;
};

export type TradeResult =
  | {
      transaction: PaperTransaction;
      new_cash_balance: number;
      holding: { instrument_id: string; quantity: number; avg_cost: number } | null;
    }
  | { error: TradeError };

export async function getPortfolio(period: "1m" | "90d" | "1y" | "all" = "90d"): Promise<PaperPortfolioResponse> {
  const res = await authFetch(`/v1/paper/portfolio?period=${period}`);
  if (!res.ok) throw new Error(`portfolio failed: ${res.status}`);
  return res.json();
}

export async function listTransactions(
  limit = 50,
  before?: string,
): Promise<{ transactions: PaperTransaction[]; has_more: boolean }> {
  const q = new URLSearchParams({ limit: String(limit) });
  if (before) q.set("before", before);
  const res = await authFetch(`/v1/paper/transactions?${q}`);
  if (!res.ok) throw new Error(`list tx failed: ${res.status}`);
  return res.json();
}

export async function trade(body: {
  instrument_id: string;
  action: "buy" | "sell";
  quantity: number;
  reason?: string;
}): Promise<TradeResult> {
  const res = await authFetch("/v1/paper/transactions", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (res.status === 422) return res.json();
  if (!res.ok) throw new Error(`trade failed: ${res.status}`);
  return res.json();
}

export async function resetPortfolio(initialCash?: number): Promise<PaperAccount> {
  const res = await authFetch("/v1/paper/reset", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(initialCash != null ? { initial_cash: initialCash } : {}),
  });
  if (!res.ok) throw new Error(`reset failed: ${res.status}`);
  return res.json();
}

export function isTradeError(r: TradeResult): r is { error: TradeError } {
  return (r as { error?: unknown }).error !== undefined;
}
